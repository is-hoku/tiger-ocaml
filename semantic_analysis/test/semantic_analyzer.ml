open Lexical_analysis
module Lexer_token = Lexical_analysis.Token
open Semantic_action
module Parser_token = Semantic_action.Token
open Semantic_analysis

let tokenize buf = Parser_token.parsertoken_of_tokent (Lexer.read buf)
let parse buf = Parser.main tokenize buf

let semant venv tenv inside_loop exp =
  Semant.trans_exp (venv, tenv, inside_loop, exp)

let test_cases =
  [
    "merge.tig";
    "queens.tig";
    "recursive_type.tig";
    "test1.tig";
    "test10.tig";
    "test11.tig";
    "test12.tig";
    "test13.tig";
    "test14.tig";
    "test15.tig";
    "test16.tig";
    "test17.tig";
    "test18.tig";
    "test19.tig";
    "test2.tig";
    "test20.tig";
    "test21.tig";
    "test22.tig";
    "test23.tig";
    "test24.tig";
    "test25.tig";
    "test26.tig";
    "test27.tig";
    "test28.tig";
    "test29.tig";
    "test3.tig";
    "test30.tig";
    "test31.tig";
    "test32.tig";
    "test33.tig";
    "test34.tig";
    "test35.tig";
    "test36.tig";
    "test37.tig";
    "test38.tig";
    "test39.tig";
    "test4.tig";
    "test40.tig";
    "test41.tig";
    "test42.tig";
    "test43.tig";
    "test44.tig";
    "test45.tig";
    "test46.tig";
    "test47.tig";
    "test48.tig";
    "test49.tig";
    "test5.tig";
    "test6.tig";
    "test7.tig";
    "test8.tig";
    "test9.tig";
  ]

let () =
  List.iter
    (fun file_name ->
      let fin = open_in (Printf.sprintf "../../example_program/%s" file_name) in
      let buf = Lexing.from_channel fin in
      try
        let ast = parse buf in
        let expty = semant Env.base_venv Env.base_tenv false ast in
        print_endline (Types.string_of_ty expty.ty);
        close_in fin
      with
      | Lexer.Syntax_error e ->
          print_endline ("SyntaxError: " ^ e);
          close_in fin
      | Parsing.Parse_error ->
          print_endline "SyntaxError";
          close_in fin
      | Error.Type_error -> close_in fin
      | Error.Name_error -> close_in fin)
    test_cases
