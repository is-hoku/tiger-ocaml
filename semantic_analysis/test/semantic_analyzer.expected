NameError: unexpected break in line 7 at character 57
error
unit
unit
unit
array of int
TypeError: expression 5 + 6 has type int but expected type unit in line 2 at character 5
error
TypeError: expression   has type string but expected type int in line 2 at character 3
error
unit
TypeError: expression df has type string but expected type int in line 3 at character 3
error
TypeError: expression rec has type {name:string,id:int} but expected type int or string in line 12 at character 10
TypeError: expression arr has type array of int but expected type int or string in line 12 at character 10
error
TypeError: expression 3 has type int but expected type unit in line 3 at character 2
error
TypeError: type c has an illegal type circulation in line 4 at character 8
TypeError: type d has an illegal type circulation in line 6 at character 8
error
NameError: type treelist is not declared in line 4 at character 30
error
NameError: function do_nothing2 is not declared in line 5 at character 14
error
NameError: variable a is not declared in line 8 at character 16
error
array of name of myint(rec)
NameError: variable i is not declared in line 3 at character 18
error
TypeError: expression nfactor(n - 1) has type unit but expected type int in line 8 at character 11
error
NameError: field nam is not declared in record in line 7 at character 5
error
TypeError: expression 3 has type int but expected type string in line 7 at character 13
error
TypeError: variable d has type int but expected type array in line 5 at character 2
error
TypeError: variable f has type int but expected type record in line 5 at character 2
error
TypeError: expression var has type string but expected type int in line 3 at character 3
error
int
TypeError: expression rec1 has type {name:string,id:int} but expected type {name:string,id:int} in line 7 at character 19
NameError: variable rec1 is not declared in line 9 at character 5
error
TypeError: expression arr1 has type array of int but expected type array of int in line 7 at character 19
NameError: variable arr1 is not declared in line 9 at character 5
error
{name:string,age:int}
int
TypeError: expression a has type string but expected type int in line 3 at character 10
NameError: variable a is not declared in line 5 at character 2
error
TypeError: expression arrayty and   has different types int and string in line 6 at character 17
TypeError: required type declared variable in line 6 at character 4
error
NameError: type rectype is not declared in line 3 at character 16
TypeError: required type declared variable in line 3 at character 4
error
TypeError: argument one has type string but expected typeint in line 5 at character 2
error
TypeError: argument one has type string but expected typeint in line 5 at character 2
TypeError: the number of arguments for function g is not correct in line 5 at character 2
error
TypeError: the number of arguments for function g is not correct in line 5 at character 2
error
int
NameError: type a is already declared in line 4 at character 5
error
NameError: function g is already declared in line 4 at character 9
error
int
TypeError: function g return type int but expected type unit in line 3 at character 9
error
int
unit
TypeError: expression a has type unit but expected type int in line 6 at character 4
error
unit
TypeError: required type declared variable in line 5 at character 4
NameError: variable a is not declared in line 7 at character 2
error
int
int
int
SyntaxError: invalid syntax in line 5 at character 20
TypeError: required type declared variable in line 5 at character 4
NameError: variable a is not declared in line 7 at character 2
error
{hd:int,tl:name of intlist(rec)}
unit
int
int
TypeError: expression 13 and   has different types int and string in line 3 at character 2
error
